### PR TITLE
Improve bundle load speeds

### PR DESCRIPTION
**Pull request by thlassche**
### Description
When a bundle has a lot of selection, the tax GetPriceConfigurationObserver becomes a real performance hog because it loads the selections over and over again. This fixes that by caching the selection in memory

### Manual testing scenarios
Create a bundle with a lot of selections. Check the loading speed of the product page with the bundle before and after this commit. In my local testing scenarios, the loading time went from ~30s to about ~3s

### CODE REVIEW

**Comment by magento-cicd2** [comment]
[CLA assistant check](https://cla.dev.magento.com/magento/magento2?pullRequest=20877) <br/>All committers have signed the CLA.

**Comment by magento-engcom-team** [comment]
Hi @thlassche. Thank you for your contribution 
Here is some useful tips how you can test your changes using Magento test environment.
Add the comment under your pull request to deploy test or vanilla Magento instance:
- `@magento-engcom-team give me test instance` - deploy test instance based on PR changes
- `@magento-engcom-team give me 2.3-develop instance` - deploy vanilla Magento instance

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation

**Comment by slavvka** [comment]
Hi @thlassche Thank you for your improvement. I would like you to clarify the following.
Is your improvement intended to cache options between the observer runs? If yes it may be unpredictable since the cached objects may be changed. If no you should clear the cache before using it since this particular observer is shared object. And this case looks rare since in most cases options of a bundle products are different objects.

**Comment by thlassche** [comment]
@slavvka I definitely agree with you, however all scenarios will hold: The observer takes the current product from registry (it already did so before my change):

```
$product = $this->registry->registry('current_product');
```
The id of that product is used as cache key for cached instances of the bundle selections, so it will not clash.

And to clarify a little bit on the potential gains of this improvement: The observer contains a recursive function (`recurConfigAndUpdatePrice`), and that one calls the `updatePriceForBundle` multiple times, which would load the same bundle selections over and over again for every call, with my improvement that is cached.

**Comment by slavvka** [comment]
> The id of that product is used as cache key for cached instances of the bundle selections, so it will not clash.

I agree but products in cache may be changed between runs of the observer and already cached products may get invalid in a next run. Since caching is needed only for the current run I would explicitly clear the cache in the beginning of `execute` method.

> And to clarify a little bit on the potential gains of this improvement: The observer contains a recursive function (`recurConfigAndUpdatePrice`), and that one calls the `updatePriceForBundle` multiple times, which would load the same bundle selections over and over again for every call, with my improvement that is cached.

Yes, now that makes sense for me. Thank you!

**Comment by thlassche** [comment]
> I agree but products in cache may be changed between runs of the observer and already cached products may get invalid in a next run. Since caching is needed only for the current run I would explicitly clear the cache in the beginning of `execute` method.

I agree, that's better. Just changed it

File: app/code/Magento/Tax/Observer/GetPriceConfigurationObserver.php [hunk 1]
```diff
@@ -23,6 +23,8 @@ class GetPriceConfigurationObserver implements ObserverInterface
      */
     protected $registry;
 
+    private $selectionCache = [];
```
Comment (slavvka) [inline app/code/Magento/Tax/Observer/GetPriceConfigurationObserver.php:32]: Please add a comment here

Comment (thlassche) [reply app/code/Magento/Tax/Observer/GetPriceConfigurationObserver.php:32]: Done

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-4085](https://jira.corp.magento.com/browse/ENGCOM-4085) has been created to process this Pull Request

**Comment by magento-engcom-team** [comment]
@thlassche thank you for contributing. Please accept Community Contributors team invitation [here](https://github.com/orgs/magento/invitation) to gain extended permissions for this repository.

File: app/code/Magento/Tax/Test/Unit/Observer/GetPriceConfigurationObserverTest.php [hunk 2]
```diff
@@ -118,7 +118,7 @@ public function testExecute($testArray, $expectedArray)
 
         $product = $this->createPartialMock(
             \Magento\Bundle\Model\Product\Type::class,
-            ['getTypeInstance', 'getTypeId', 'getStoreId', 'getSelectionsCollection']
+            ['getTypeInstance', 'getTypeId', 'getStoreId', 'getSelectionsCollection','getId']
```
Comment (slavvka) [inline app/code/Magento/Tax/Test/Unit/Observer/GetPriceConfigurationObserverTest.php]: Hey @thlassche you missed the space

**Review by slavvka — DISMISSED** [review]

**Review by slavvka — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @slavvka, thank you for the review.
[ENGCOM-4085](https://jira.corp.magento.com/browse/ENGCOM-4085) has been created to process this Pull Request

**Review by sidolov — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @sidolov, thank you for the review.
[ENGCOM-4085](https://jira.corp.magento.com/browse/ENGCOM-4085) has been created to process this Pull Request

**Comment by engcom-Delta** [comment]
Hi, @thlassche . Your changes in code does't do anything because my debugger don't stop at line 116 when i load a page of  bundle product. And this page load time like in 2.3-develop. 
Could you take a look? 
Thanks!

**Comment by thlassche** [comment]
Hi @engcom-Delta,

Just tried here with a `debug_print_backtrace` 116, which gives:

```
#0  Magento\Tax\Observer\GetPriceConfigurationObserver->updatePriceForBundle() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-tax/Observer/GetPriceConfigurationObserver.php:90]
--
  | #1  Magento\Tax\Observer\GetPriceConfigurationObserver->recurConfigAndUpdatePrice() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-tax/Observer/GetPriceConfigurationObserver.php:86]
  | #2  Magento\Tax\Observer\GetPriceConfigurationObserver->recurConfigAndUpdatePrice() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-tax/Observer/GetPriceConfigurationObserver.php:86]
  | #3  Magento\Tax\Observer\GetPriceConfigurationObserver->recurConfigAndUpdatePrice() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-tax/Observer/GetPriceConfigurationObserver.php:86]
  | #4  Magento\Tax\Observer\GetPriceConfigurationObserver->recurConfigAndUpdatePrice() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-tax/Observer/GetPriceConfigurationObserver.php:86]
  | #5  Magento\Tax\Observer\GetPriceConfigurationObserver->recurConfigAndUpdatePrice() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-tax/Observer/GetPriceConfigurationObserver.php:61]
  | #6  Magento\Tax\Observer\GetPriceConfigurationObserver->execute() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/framework/Event/Invoker/InvokerDefault.php:72]
  | #7  Magento\Framework\Event\Invoker\InvokerDefault->_callObserverMethod() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/framework/Event/Invoker/InvokerDefault.php:60]
  | #8  Magento\Framework\Event\Invoker\InvokerDefault->dispatch() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/framework/Event/Manager.php:66]
  | #9  Magento\Framework\Event\Manager->dispatch() called at [/Users/thlassche/Projects/heuveling/httpdocs/generated/code/Magento/Framework/Event/Manager/Proxy.php:95]
  | #10 Magento\Framework\Event\Manager\Proxy->dispatch() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/module-bundle/Block/Catalog/Product/View/Type/Bundle.php:210]
  | #11 Magento\Bundle\Block\Catalog\Product\View\Type\Bundle->getJsonConfig() called at [/Users/thlassche/Projects/heuveling/httpdocs/vendor/magento/framework/Interception/Interceptor.php:58]

```

Is your Magento configured to display prices including tax?

**Comment by engcom-Delta** [comment]
@thlassche , i test with display prices including tax. Sorry, but i can't test with debug_print_backtrace because i got this message when i try to clone this PR: "Remote branch EENGCOM-4085-magento-magento2-20877 not found in upstream origin". We have to create new PR.

**Comment by engcom-Delta** [comment]
:heavy_check_mark: QA passed

**Pull request merged**